lib/msf/core/payload.rb

```diff
@@ -644,7 +645,7 @@ def build(asm, off={})
     end
 
     # Assemble the payload from the assembly
-    a = self.arch
+    a = opts[:arch] || self.arch
```

Comment:
It looks like the original adapted arch is passed i from `opts`, so this is:

```
[1] pry(#<#<Class:0x00007fab4c96f308>>)> opts
=> {:arch=>"x86"}
```

But I also had this line as this previously, similar to the pattern over [here](https://github.com/rapid7/metasploit-framework/blob/453c8d8d6986934a3b869733fa1d01eaaa1fbbb0/lib/msf/util/payload_cached_size.rb#L142-L145)

```
    a = module_info["AdaptedArch"] || self.arch
```

From a quick look, it seems like there's other references in this file too that use `self.arch`, which might also be wrong

Does anyone have any preferences on the chosen approach here?

Reply:
I think the way it is currently is correct. `self.arch` is the architecture of the over all payload which would be ARCH_CMD in the case of `cmd/windows/powershell/download_exec` but at this point, it needs to be ARCH_X86 as specified in the options from the caller. This is the pattern I took in #16579. :+1: